src/hopen/core.cljc

```diff
@@ -19,25 +19,70 @@
             :else element))]
    (f-eval element)))
 
+;; Top level evaluation functions
+;; The given element can be a "block" to be executed or an simple expression to be evaluated
 (defn- rf-block [rf env result element]
   (or (when (list? element)
         (let [[f-symb & args] element
-              f (get-in env [:block-macro f-symb])]
-          (when f
+              f (get-in env [:block-macro f-symb])
+
+              ;; Values in :block-macro could be a function or a var
+              ;; If it's a var, resolve it now (into a function).
+              ;; This allows us to frequently rebind/re-eval functions without
+              ;; having to reconstruct `default-env`
```

Comment:
I would prefer to avoid having that in the code.

There are some ways, to solve this practical thing differently. In IntellijIDEA, I am reloading the whole file with a shortcut (Cmd+Shift+L).

Reply:
How does intellij deal with line level debugging? Does it need a separate instrumentation step?

I'm running spacemacs + cider. When I need to do line level debugging, the function needs to be re-evaluated and (presumably) have some instrumentation inserted. This means... even if I re-eval the whole file, I'd still have to re-instrument the function I want to debug. What then happens is that `default-env` is still holding on to the slightly older version of the function, causing the debugger to not catch the first invocation/iteration.

Lastly, I'll just add that while a line level debugger is often not required when working in clojure, because of the repl, it really is invaluable when working with recursive algorithms that works on trees of data.

Reply:
I don't use line-level debugging. Now I understand your problem.

I do debug testing at the function level, however. I use:
- `conj` for `rf`
- `[]` for `result`,
- `{}` for `env`,

When I know that the function is working, I add tests for the transducer which uses the function and I commit those.

Reply:
My other method of debugging .. 
![image](https://user-images.githubusercontent.com/598193/55101117-5cc76f00-50fe-11e9-90fb-4a59a3a2abfd.png)